File: ruby/exercises/beer-song/beer_song.rb
# Beer Song
class BeerSong
  def verses(start = 99, stop = 0)
    start.downto(stop).map do |verse_num|
      verse(verse_num)
    end.join("\n") + "\n"
  end
  alias sing verses

  def verse(num)
    [prefix(num), suffix(num)].join("\n") + "\n"
  end

  private

  def prefix(num)
    "#{quantity(num).capitalize} #{container(num)} of beer on the wall, "\
    "#{quantity(num)} #{container(num)} of beer."
  end

  def suffix(num)
    return last_line if num == 0
    "Take #{cardinality(num)} down and pass it around, #{quantity(num - 1)} "\
      "#{container(num - 1)} of beer on the wall."
  end

  def last_line
    'Go to the store and buy some more, 99 bottles of beer on the wall.'
  end

  def container(num)
    num == 1 ? 'bottle' : 'bottles'
  end

  def quantity(num)
    num == 0 ? 'no more' : num.to_s
  end

  def cardinality(num)
    num == 1 ? 'it' : 'one'
  end
end

File: go/beer-song/beer_song.go
package beer

import "fmt"

func Song() string {
	result, err := Verses(99, 0)
	if err != nil {
		panic(err)
	}
	return result
}

func Verses(start, stop int) (output string, err error) {
	if stop < 0 {
		return "", fmt.Errorf("invalid: stop verse number %d can not be negative", stop)
	}
	if start < stop {
		return "", fmt.Errorf("invalid: start %v less than stop %v", start, stop)
	}
	for i := start; i >= stop; i-- {
		verse, err := Verse(i)
		if err != nil {
			return "", err
		}
		output += verse
		output += "\n"
	}
	return output, nil
}

func Verse(n int) (string, error) {
	if n > 99 {
		return "", fmt.Errorf("invalid verse number: %d", n)
	}
	switch n {
	case 0:
		return "No more bottles of beer on the wall, no more bottles of beer.\nGo to the store and buy some more, 99 bottles of beer on the wall.\n", nil
	case 1:
		return "1 bottle of beer on the wall, 1 bottle of beer.\nTake it down and pass it around, no more bottles of beer on the wall.\n", nil
	case 2:
		return "2 bottles of beer on the wall, 2 bottles of beer.\nTake one down and pass it around, 1 bottle of beer on the wall.\n", nil
	default:
		return fmt.Sprintf("%d bottles of beer on the wall, %d bottles of beer.\nTake one down and pass it around, %d bottles of beer on the wall.\n", n, n, n-1), nil
	}
}
